src/parser.rs

```diff
@@ -198,10 +199,94 @@ fn parse_unknown(input: (&[u8], usize)) -> IResult<(&[u8], usize), MessageKind>
     Ok((input, MessageKind::Unknown))
 }
 
+pub fn decode_id_13_field(f: u16) -> u16 {
+    let mut hex_gillham = 0;
+    if f & 0x1000 != 0 {
+        hex_gillham |= 0x0010;
+    } // Bit 12 = C1
+    if f & 0x0800 != 0 {
+        hex_gillham |= 0x1000;
+    } // Bit 11 = A1
+    if f & 0x0400 != 0 {
+        hex_gillham |= 0x0020;
+    } // Bit 10 = C2
+    if f & 0x0200 != 0 {
+        hex_gillham |= 0x2000;
+    } // Bit  9 = A2
+    if f & 0x0100 != 0 {
+        hex_gillham |= 0x0040;
+    } // Bit  8 = C4
+    if f & 0x0080 != 0 {
+        hex_gillham |= 0x4000;
+    } // Bit  7 = A4
+      //if (ID13Field & 0x0040) {hexGillham |= 0x0800;} // Bit  6 = X  or M
+    if f & 0x0020 != 0 {
+        hex_gillham |= 0x0100;
+    } // Bit  5 = B1
+    if f & 0x0010 != 0 {
+        hex_gillham |= 0x0001;
+    } // Bit  4 = D1 or Q
+    if f & 0x0008 != 0 {
+        hex_gillham |= 0x0200;
+    } // Bit  3 = B2
+    if f & 0x0004 != 0 {
+        hex_gillham |= 0x0002;
+    } // Bit  2 = D2
+    if f & 0x0002 != 0 {
+        hex_gillham |= 0x0400;
+    } // Bit  1 = B4
+    if f & 0x0001 != 0 {
+        hex_gillham |= 0x0004;
+    } // Bit  0 = D4
+
+    hex_gillham
+}
+
+fn parse_surveillance_identity(input: (&[u8], usize)) -> IResult<(&[u8], usize), ModeSMessageKind> {
+    let (_input, (_df, _flight_status, _downlink_req, _utility_msg, id_code, _parity)): (
+        _,
+        (u8, u8, u8, u8, u16, u32),
+    ) = tuple((
+        tag_bits(0b00101, 5u8),
+        take_bits(3u8),
+        take_bits(5u8),
+        take_bits(6u8),
+        take_bits(13u8),
+        take_bits(24u8),
+    ))(input)?;
+    let squawk_code = decode_id_13_field(id_code);
+    Ok((
+        input,
+        ModeSMessageKind::SurveillanceIdentity {
+            squawk: Squawk::from_u16(squawk_code),
+        },
+    ))
+}
+
+fn parse_mode_s_message_kind(input: (&[u8], usize)) -> IResult<(&[u8], usize), ModeSMessageKind> {
+    parse_surveillance_identity(input)
+}
+
+fn parse_mode_s_message(input: (&[u8], usize)) -> IResult<(&[u8], usize), MessageKind> {
+    let (input, kind) = parse_mode_s_message_kind(input)?;
+    let crc = mode_s_crc(input.0, 7).unwrap();
```

Comment:
This shouldn't panic. Unfortunately I haven't found a good way to handle custom parsing errors using nom so something like `.map_err(|| Err::Error(make_error(input, ErrorKind::LengthValue)))?;`, while not ideal,  would at least avoid crashing here.